### PR TITLE
MAKE-1293: implement archive stream bucket

### PR DESCRIPTION
**Pull request by Kimchelly**
JIRA: https://jira.mongodb.org/browse/MAKE-1293

~This is a WIP. The Push/Pull functionality works, but is very special case currently. I had to skip a lot of tests that use `List()` because a lot of the bucket interface doesn't map well to a single archived file in S3 since the working assumption of the other buckets is that 1 object in S3 = 1 file in the local file system. Also, manipulating the archive to maintain the file system structure is undesirable to handle single-file cases like `Get()`/`Put()`. We could just make an interface that has `Push()`/`Pull()` if we want to enforce that only `Push()`/`Pull()` gets used for this special bucket.~ I moved the archive bucket to its own special interface and wrote tests.

* Implement Push/Pull based on archive tar stream. ~The other methods depend on the multipart bucket and are mostly for the sake of satisfying the interface. Really, you should probably only use this implementation for `Push()`/`Pull()` and nothing else.~
* Write tests where possible. ~Will write more archive-specific tests for push/pull later.~

### CODE REVIEW

File: util.go [hunk 1]
```diff
@@ -181,3 +183,174 @@ func deleteOnPull(ctx context.Context, sourceFiles []string, local string) error
 
 	return catcher.Resolve()
 }
+
+// The archive/unarchive functions below are modified version of the same
+// functions from github.com/mholt/archiver.
+
+func tarFile(tarWriter *tar.Writer, dir, relPath string) error {
```
Comment (tychoish) [inline util.go:190]: should the utility stuff end up in the utility package?

Comment (Kimchelly) [reply util.go:190]: I don't have strong feelings about it going here or in utility, but I feel like there's not a strong use case for this function. Most use cases could probably be satisfied by the functionality in archiver. I just happened to need to slightly tweak archiver's algorithm to make it work more nicely with pail buckets.

Comment (tychoish) [reply util.go:190]: got it.

File: util.go [hunk 2]
```diff
@@ -181,3 +183,174 @@ func deleteOnPull(ctx context.Context, sourceFiles []string, local string) error
 
 	return catcher.Resolve()
 }
+
+// The archive/unarchive functions below are modified version of the same
```
Comment (julianedwards) [inline util.go:187]: versions

File: s3_bucket_util_test.go [hunk 3]
```diff
@@ -589,6 +589,29 @@ func getS3LargeBucketTests(ctx context.Context, tempdir, s3BucketName, s3Prefix,
 	}
 }
 
+// TODO: write archive-specific tests for push/pull.
```
Comment (julianedwards) [inline s3_bucket_util_test.go:592]: are you going to do this here? or in another ticket/PR?

Comment (Kimchelly) [reply s3_bucket_util_test.go]: I actually already wrote them below in `TestS3ArchiveBucket`. I removed these lines.

File: util.go (hunk 2 above)
Comment (Kimchelly) [reply util.go]: 👍

**Pull request merged**